model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final GiftCertificateDao certificateDao;
+    private final GiftCertificateDTOMapper dtoMapper;
+
+    public GiftCertificateServiceImpl(GiftCertificateDao certificateDao, GiftCertificateDTOMapper dtoMapper) {
+        this.certificateDao = certificateDao;
+        this.dtoMapper = dtoMapper;
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO find(Long id) throws ServiceException, NotExistEntityException {
+        try {
+            Optional<GiftCertificate> certificate = certificateDao.findById(id);
+            return certificate.map(dtoMapper::toDTO)
+                    .orElseThrow(() -> new NotExistEntityException("Gift certificate with id=" + id + " not exist!"));
+        } catch (DataAccessException e) {
+            log.error("Find certificate service exception", e);
+            throw new ServiceException("Find certificate service exception", e);
+        }
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO add(CertificateDTO certificateDTO) throws ServiceException {
+        try {
+            GiftCertificate certificate = dtoMapper.fromDTO(certificateDTO);
+            GiftCertificate giftCertificate = certificateDao.create(certificate);
+            return dtoMapper.toDTO(giftCertificate);
+        } catch (DataAccessException e) {
+            log.error("Add certificate service exception", e);
+            throw new ServiceException("Add certificate service exception", e);
+        }
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO update(CertificateDTO certificateDTO) throws ServiceException {
+        try {
+            GiftCertificate certificate = dtoMapper.fromDTO(certificateDTO);
+            GiftCertificate updated = certificateDao.update(certificate);
+            return dtoMapper.toDTO(updated);
+        } catch (DataAccessException e) {
+            log.error("Update certificate service exception", e);
+            throw new ServiceException("Update certificate service exception", e);
+        }
+    }
+
+    @Override
+    @Transactional
+    public void delete(Long id) throws ServiceException {
+        try {
+            certificateDao.delete(id);
+        } catch (DataAccessException e) {
+            log.error("Delete certificate service exception", e);
+            throw new ServiceException("Delete certificate service exception", e);
+        }
+    }
+
+    @Override
+    @Transactional
+    public List<CertificateDTO> filterByParameters(
+            String tag, String part, String sortBy, SortType type, int page, int size)
+            throws ServiceException, IncorrectArgumentException {
+        try {
+            long count = count();
+            Page certificatePage = new Page(page, size, count);
+            List<String> tags = ParameterParser.parseWithAndCondition(tag);
```

Comment:
I think you should not parse it in the service layer.
And I really believe that your controller will be able to parse this tags list for you (e.g. https://medium.com/@AADota/spring-passing-list-and-array-of-values-as-url-parameters-1ed9bbdf0cb2 )

Reply:
fix